### PR TITLE
Add documents, engine, tag helpers to snapshot

### PR DESCRIPTION
**Pull request by rynowak**
The project snapshot now maintains a RazorProjectEngine as well as set
of Tag Helpers that are known for that snapshot.

Pivoted some more services to be snapshot-centric.

Also added the ability to track .cshtml documents to the project system.
For now most components just ignore document changes.

### CODE REVIEW

**Comment by rynowak** [comment]
/cc @NTaylorMullen - please take a look, these are the changes we discussed yesterday that start tracking documents through the PSM. 

Additionally we no longer use a 'work queue' - tag helpers now hang off of the snapshot and can be computed on demand. This responsibility is moved to the document tracker.

File: test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/ProjectSystem/DefaultDocumentSnapshotTest.cs [hunk 1]
```diff
@@ -1,10 +1,9 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
-namespace Microsoft.CodeAnalysis.Remote.Razor
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
 {
-    internal class GeneratedDocument
+    class DefaultDocumentSnapshotTest
```
Comment (NTaylorMullen) [inline test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/ProjectSystem/DefaultDocumentSnapshotTest.cs]: ?

File: test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/ProjectSystem/ProjectStateTest.cs [hunk 2]
```diff
@@ -0,0 +1,359 @@
...
+                .AddHostDocument(Documents[1]);
+
+            // Act
+            var state = original.AddHostDocument(new HostDocument(Documents[1].FilePath, "SomePath.cshtml"));
+
+            // Assert
+            Assert.Same(original, state);
+        }
+
+                [Fact]
```
Comment (NTaylorMullen) [inline test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/ProjectSystem/ProjectStateTest.cs]: 👀

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs [hunk 3]
```diff
@@ -89,43 +79,109 @@ public override IReadOnlyList<ProjectSnapshot> Projects
...
+                }
+
+                return projects;
             }
         }
 
         public override Workspace Workspace { get; }
 
-        public override void ProjectUpdated(ProjectSnapshotUpdateContext update)
+        public override ProjectSnapshot GetLoadedProject(string filePath)
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotManager.cs]: Declined, this is very straightforward as it is

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultDocumentSnapshot.cs [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        public override string TargetPath => State.HostDocument.TargetPath;
+
+        public override Task<RazorCodeDocument> GetGeneratedOutputAsync()
+        {
+            // IMPORTANT: Don't put more code here. We want this to return a cached task.
+            return State.GeneratedOutput.GetGeneratedOutputInitializationTask(Project, this);
+        }
+
+        public override bool TryGetGeneratedOutput(out RazorCodeDocument results)
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultDocumentSnapshot.cs]: This is here so you can get the generated assets if they are avilalable but without blocking.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/HostDocument.cs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+                throw new ArgumentNullException(nameof(targetPath));
+            }
+
+            FilePath = filePath;
+            TargetPath = targetPath;
+        }
+
+        public string FilePath { get; }
+
+        public string TargetPath { get; }
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/HostDocument.cs]: TargetPath is an MSBuildism. It's a lot like the view engine path

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectState.cs [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+        public VersionStamp Version { get; }
+
+        // Computed State
+        public ProjectEngineTracker ProjectEngine
+        {
+            get
+            {
+                if (_projectEngine == null)
+                {
+                    lock (_lock)
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/ProjectState.cs]: yeah this is just double-checked locking. I would normally use LazyInitializer for this, but I don't want the allocation for the delegate.

The alternative would be to not use a lock and use swap.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshot.cs [hunk 7]
```diff
@@ -3,182 +3,82 @@
...
         }
 
-        public bool HaveTagHelpersChanged(ProjectSnapshot original)
+        public override bool TryGetTagHelpers(out IReadOnlyList<TagHelperDescriptor> results)
         {
-            if (original == null)
+            if (State.TagHelpers.IsResultAvailable)
             {
-                throw new ArgumentNullException(nameof(original));
+                results = State.TagHelpers.GetTagHelperInitializationTask(this).Result;
```
[the remark this answers is not in the document]
Comment (rynowak) [reply src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshot.cs]: We already know the result is available.

File: test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/ProjectSystem/DefaultDocumentSnapshotTest.cs (hunk 1 above)
Comment (rynowak) [reply test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/ProjectSystem/DefaultDocumentSnapshotTest.cs]: Will remove

File: test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/ProjectSystem/ProjectStateTest.cs (hunk 2 above)
Comment (rynowak) [reply test/Microsoft.CodeAnalysis.Razor.Workspaces.Test/ProjectSystem/ProjectStateTest.cs]: lol

**Pull request merged**